src/pystack/_pystack/mem.cpp

```diff
@@ -155,36 +156,137 @@ MemoryMapInformation::setHeap(const VirtualMap& heap)
+{
+    size_t value_size = value.size();
+
+    if (!can_fit(value_size)) {
+        return;
+    }
+
+    auto it = d_cache.find(key);
+
+    if (it != d_cache.end()) {
+        d_cache_list.erase(it->second.second);
+        d_cache.erase(it);
+    }
+
+    while (d_size + value_size > d_cache_capacity) {
+        auto last = d_cache_list.rbegin();
+        d_cache.erase(last->first);
+        d_size -= last->second;
+        d_cache_list.pop_back();
+    }
+
+    d_cache_list.push_front(list_node_t{key, value_size});
+    d_cache[key] = cache_value_t{std::move(value), d_cache_list.begin()};
+    d_size += value_size;
+}
+
+const std::vector<char>&
+LRUCache::get(const uintptr_t& key)
+{
+    auto it = d_cache.find(key);
+    if (it == d_cache.end()) {
+        throw std::range_error("There is no such key in cache");
+    } else {
+        auto node_it = it->second.second;
+        d_cache_list.splice(d_cache_list.begin(), d_cache_list, node_it);
+        return it->second.first;
+    }
+}
+
+bool
+LRUCache::exists(const uintptr_t& key)
+{
+    return (d_cache.find(key) != d_cache.end());
+}
+
+bool
+LRUCache::can_fit(size_t size)
+{
+    return d_cache_capacity > size;
+}
+
+ProcessMemoryManager::ProcessMemoryManager(pid_t pid, std::vector<VirtualMap>& vmaps)
+: d_pid(pid)
+, d_vmaps(vmaps)
+, d_lru_cache(CACHE_CAPACITY)
+{
+}
+
 ProcessMemoryManager::ProcessMemoryManager(pid_t pid)
-: d_pid(pid){};
+: d_pid(pid)
+, d_lru_cache(CACHE_CAPACITY)
+{
+}
 
 ssize_t
-ProcessMemoryManager::copyMemoryFromProcess(remote_addr_t addr, size_t len, void* buf) const
+ProcessMemoryManager::readChunk(remote_addr_t addr, size_t len, char* dst) const
 {
     struct iovec local[1];
     struct iovec remote[1];
-
-    local[0].iov_base = buf;
-    local[0].iov_len = len;
-    remote[0].iov_base = (void*)addr;
-    remote[0].iov_len = len;
-
-    ssize_t result = _process_vm_readv(d_pid, local, 1, remote, 1, 0);
-    if (result < 0) {
-        if (errno == EFAULT) {
-            throw InvalidRemoteAddress();
-        } else if (errno == EPERM) {
-            throw std::runtime_error(PERM_MESSAGE);
+    ssize_t result = 0;
+    ssize_t read = 0;
+
+    do {
+        local[0].iov_base = dst + result;
+        local[0].iov_len = len - result;
+        remote[0].iov_base = reinterpret_cast<uint8_t*>(addr) + result;
+        remote[0].iov_len = len - result;
+
+        read = _process_vm_readv(d_pid, local, 1, remote, 1, 0);
+        if (read < 0) {
+            if (errno == EFAULT) {
+                throw InvalidRemoteAddress();
+            } else if (errno == EPERM) {
+                throw std::runtime_error(PERM_MESSAGE);
+            }
+            throw std::system_error(errno, std::generic_category());
         }
-        throw std::system_error(errno, std::generic_category());
-    }
+
+        result += read;
+    } while ((size_t)read != local[0].iov_len);
 
     if ((size_t)result != len) {
         throw InvalidCopiedMemory();
     }
```

Comment:
Unless I'm missing something, it doesn't seem like this can happen anymore - if we've broken out of the `while` loop, it's because `read == local[0].iov_len`, and `local[0].iov_len` was the amount we needed to read before `result` would equal `len`, so if `read == local[0].iov_len`, then `result == len`.

```suggestion
```